tools/shared/src/main/scala/org/scalajs/core/tools/sem/Semantics.scala

```diff
@@ -107,21 +119,124 @@ final class Semantics private (
       moduleInit: CheckedBehavior = this.moduleInit,
       strictFloats: Boolean = this.strictFloats,
       productionMode: Boolean = this.productionMode,
-      runtimeClassName: RuntimeClassNameFunction = this.runtimeClassName): Semantics = {
+      runtimeClassNameMapper: RuntimeClassNameMapper =
+        this.runtimeClassNameMapper): Semantics = {
     new Semantics(
         asInstanceOfs = asInstanceOfs,
         arrayIndexOutOfBounds = arrayIndexOutOfBounds,
         moduleInit = moduleInit,
         strictFloats = strictFloats,
         productionMode = productionMode,
-        runtimeClassName = runtimeClassName)
+        runtimeClassNameMapper = runtimeClassNameMapper)
   }
 }
 
 object Semantics {
   private val HashSeed =
     scala.util.hashing.MurmurHash3.stringHash(classOf[Semantics].getName)
 
+  sealed abstract class RuntimeClassNameMapper {
+    import RuntimeClassNameMapper._
+
+    def andThen(that: RuntimeClassNameMapper): RuntimeClassNameMapper = {
```

Comment:
I guess this is fine for now, but at some point people might want the ability to "abort" the matching chain (as in: do these replacements, then drop the rest). If this is requested, we can add a `orElse` method I assume (and adjust internals).